Feature/4 create entity

42jeelee opened the pull request:
### Comment 와 Category Entity 추가
- CommentEntity
  + id
  + author
    + 기본 `String` 타입입니다.
  + content
  + post
    + PostEntity의 외래키, 해당 PostEntity가 삭제될 시 같이 삭제됨
  
- CategoryEntity
  + id
  + name

### PostEntity.create -> PostEntity.of
다른 클래스와 통일성을 위해 변경하였습니다.

42jeelee committed:
feat: Add field views at post entity
when getPostById increase views

Is this annotation method ok?

Commit:
Feat: Add eventPublisher
increaseViewsById works in an async event manner.

- As events increase, they need to be structured differently than @Async. (maybe. SimpleApplicationEventMulticaster)
- we need use recode keyword

Commit:
Feat: Add Entity for Comment and Category
- add - add JPARepository for Comment and Category
- style - PostEntity.create -> PostEntity.of

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/comment/entity/CommentEntity.java:
```diff
@@ -0,0 +1,43 @@
+	@Column(nullable = false)
+	private String content;
+
+	@ManyToOne(cascade = CascadeType.REMOVE)
```
CommentEntity -> PostEntity 참조 보다는 
반대로 PostEntity -> CommentEntity 참조가 더 좋을 것 같아요.

게시글을 불러올 때 코멘트를 같이 불러오는 케이스는 있겠지만 CommentEntity를 불러오면서 게시글을 불러오는 케이스는 없을 것 같거든요.

cascade 옵션에는 어떤 것들이 있고, 각각이 어떤 동작을 하는지 잘 알고 만드는 것이 좋습니다.
각가의 옵션을 나열하고 동작을 설명해 보세요. 

그리고 @ManyToOne(cascade = CascadeType.REMOVE) 으로 설정 했을 경우 CommentEntity 가 삭제되었을 때 어떻게 동작하는지 설명해 보세요.

42jeelee replied:
### JPA cascadeType

- `CascadeType.ALL` : 아래의 모든 옵션을 포함
- `CascadeType.PERSIST` : 영속화할 때
- `CascadeType.MERGE` : 병합할 때
- `CascadeType.REMOVE` : 삭제할 때
- `CascadeType.REFLASH` : 새로고침할 때
- `CascadeType.DETACH` : 분리할 때

cascade 옵션은 총 6가지가 있습니다! 기본적으로 `cascade`는 위와 같은 5가지 동작(ALL은 제외)과 같은 특정 행위를 할 때, 연관된 엔티티도 같은 동작을 취하게끔 JPA에서 지원하는 옵션입니다.

```java
@manyToOne(cascade = CascadeType.REMOVE)
private PostEntity post;
```
만약, 위와 같이 설정했을 경우 해당 Comment 엔티티를 `remove()`할 때 참조된 Post 엔티티도 같이 `remove()` 됩니다. 

#### 해당 코드는 구현하지 않고 정의만 해서 제대로 확인하지 못한 점 죄송합니다 ..
다음 **PR #7** 이 구현파트이며 구현하는 과정에서 의도된 바와 반대로 됬다는 사실을 깨닫고 수정하였습니다 ..

수정된 커밋 : [c5442b8](https://github.com/42jeelee/demoBoard-server/pull/7/commits/c5442b8d9c403b0e98cdd0f50223b01ba32fb94a)

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/post/listener/PostEventListener.java:
```diff
@@ -0,0 +1,23 @@
+public class PostEventListener {
+	private final PostService postService;
+
+	@Async
```
@Async 를 너무 남발 하는 것 같아요. service 에만 달거나 EventListener 달거나 둘중에 하나만 달아도 충분해 보입니다.

rarelhw reviewed (commented):
리뷰 확인 부탁드립니다.

42jeelee committed:
FIX: Fix JPA cascade options
- Add Comments at PostEntity
- Add Posts at CategoryEntity
- remove async at PostService

- Need handle HttpMessageNotReadableException
- record keyword

42jeelee commented:
### 수정사항을 반영하여 커밋하였습니다.
- PostEntity - CommentEntity 양방향 참조
  + PostEntity -> CommentEntity : cascade를 위해 추가해주었습니다. ( `REMOVE` 시 해당 댓글도 삭제 )
  + CommentEntity -> PostEntity : **[ 게시글 존재유무 ]** 와 **[ 삭제 시 해당 댓글이 게시글의 댓글인지 확인  ]** 하는 로직에 필요하므로 추가하였습니다.
  > 아직 인증/인가 부분이 없기 때문에 무분별한 삭제를 방지하기 위한 최소한의 조치 ..입니다

- Async 중복 제거
  + PostService의 Async를 제거해주었습니다.

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/category/entity/CategoryEntity.java:
```diff
@@ -0,0 +1,24 @@
+	@Column(nullable = false)
+	private String name;
+
+	@OneToMany(mappedBy = "category", cascade = CascadeType.REMOVE)
```
카테고리가 지워지면 post도 같이 지워져야 하는 스팩인가요? 요건 조금 위험할꺼 같은데.. post -> category 로 가는 참조는 OK

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/post/dto/request/PostCreateRequest.java:
```diff
@@ -23,5 +20,9 @@ public class PostCreateRequest {
 	@Pattern(regexp = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[\\W_]).+$", message = "Password must contain at least one number/letter/special character.")
 	private final String password;
 
+	@NotNull(message = "Category cannot be NULL.")
```
ㅎㅎ 메시지 정의도 좋네요. 나중에 i18n도 해서 다국어 처리도 해보면 좋겠어요 ^^

rarelhw commented on src/main/java/kr/co/jeelee/demoboard/domain/post/entity/PostEntity.java:
```diff
@@ -25,18 +25,30 @@ public class PostEntity extends BaseTimeEntity {
 
 	private String password;
 
+	@ManyToOne
+	@JoinColumn(nullable = false)
+	private CategoryEntity category;
+
+	@OneToMany(mappedBy = "post", cascade = CascadeType.REMOVE)
+	private List<CommentEntity> comments;
```
OK 관계 설정 좋네요.

rarelhw reviewed (approved):
좋아용! 개발중에 풀리퀘 너무 길어지는 것도 힘드니까 얘는 머지 하시죵!
